### PR TITLE
[SPARK-15362] [ML] Make spark.ml KMeansModel load backwards compatible

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?

[SPARK-14646](https://issues.apache.org/jira/browse/SPARK-14646) makes `KMeansModel` store the cluster centers one per row. `KMeansModel.load()` method needs to be updated in order to load models saved with Spark 1.6.
## How was this patch tested?

Since `save/load` is `Experimental` for 1.6, I think offline test for backwards compatibility is enough.

### CODE REVIEW

**Comment by yanboliang** [comment]
cc @jkbradley @mengxr

File: mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala [hunk 1]
```diff
@@ -206,11 +212,21 @@ object KMeansModel extends MLReadable[KMeansModel] {
...
-
       val dataPath = new Path(path, "data").toString
-      val data: Dataset[Data] = sqlContext.read.parquet(dataPath).as[Data]
-      val clusterCenters = data.collect().sortBy(_.clusterIdx).map(_.clusterCenter)
-      val model = new KMeansModel(metadata.uid, new MLlibKMeansModel(clusterCenters))
+
+      val versionRegex = "([0-9]+)\\.(.+)".r
+      val versionRegex(major, _) = metadata.sparkVersion
+
+      val model = if (major.toInt >= 2) {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala]: I think you can just test via ```metadata.sparkVersion >= "2.0"```

File: mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala [hunk 2]
```diff
@@ -206,11 +212,21 @@ object KMeansModel extends MLReadable[KMeansModel] {
...
+        val data: Dataset[Data] = sqlContext.read.parquet(dataPath).as[Data]
+        val clusterCenters = data.collect().sortBy(_.clusterIdx).map(_.clusterCenter)
+        new KMeansModel(metadata.uid, new MLlibKMeansModel(clusterCenters))
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala]: Factor this line out to avoid duplication

**Comment by jkbradley** [comment]
Thanks!  I should have done this before.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala (hunk 1 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala]: @jkbradley It does not work as expected when spark version > 10.0.
For example, if ```metadata.sparkVersion = "10.0"```, ```metadata.sparkVersion >= "2.0"``` will return false which is not correct.
We can not use ```metadata.sparkVersion.toDouble``` to compare as well. For example, if ```metadata.sparkVersion = "1.6.1"```, ```metadata.sparkVersion.toDouble``` will throw the following  exception
```
java.lang.NumberFormatException: multiple points
```
So I think using regex is the most robust method.

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala]: Good point...we might have to fix this in some other places in Spark once we get to version 10!

Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala]: I searched the code base and found that only ```PCAModel.load()``` use ```sparkVersion``` comparison and it was implemented correctly.

**Comment by jkbradley** [comment]
LGTM
Merging with master and branch-2.0
Thanks!

**Pull request closed without merging**